wagtail/core/collectors.py

```diff
@@ -0,0 +1,722 @@
+                yield use
+
+    @classmethod
+    def from_nested_list(cls, nested_list, parent=None, on_delete=CASCADE,
+                         exclude=None, originals=()):
+        """
+        Given a sequence of model instances, which may contain sub-lists indicating cascading deletions,
+        yield a sequence of Use records.
+
+        ``nested_list`` is a complicated data structure returned by ``NestedObjects.nested()``.
+        The root contains all the objects passed in ``NestedObjects, as well as some related objects.
+        Suppose we have multiple items:
+        - ``original1`` & ``original2``, the originals we want to delete
+        - ``m2m11`` & ``m2m12``, two many-to-many items from ``original1``
+        - ``m2m21``, a many-to-many item from ``original2``
+        - ``related_fk``, an item that has a foreign key pointing to ``origina1``
+        - ``one_to_one``, an item that has a one_to_one pointing from or to ``original1``
+        - ``related_gfk``, an item that has a generic foreign key pointing to ``original1``
+        Then ``nested_list`` will be:
+        [original1, [m2m11, m2m12, related_fk], original2, [m2m21], one_to_one, related_gfk]
+
+        :param nested_list: Iterable sequence of model instances; may contain sub-lists which indicate
+            cascading deletions from the previous model instance in the sequence
+        :param parent: For cascading deletions, the Use record to assign as the parent
+        :param on_delete: on_delete behaviour to assign to the Use records
+        :param exclude: Set of Use records to omit from the returned sequence. Will be modified
+            in-place to include all returned records
+        :param originals: List of original instances which we're finding references to, as opposed to ones which
+            appear in ``nested_list`` as part of a CASCADE chain
+        """
+        if exclude is None:
+            exclude = set()
+
+        # Retain the Use record for the most-recently-handled instance, as this will be used as the parent
+        # for any sub-lists we encounter
+        parent_use = None
+
+        for obj in nested_list:
+            if isinstance(obj, list):
+                yield from cls.from_nested_list(
+                    obj, parent=parent_use, on_delete=on_delete, exclude=exclude)
+            else:
+                use = cls(obj, parent=parent, on_delete=on_delete)
+                # When ``use`` was already listed, we don’t want to list it
+                # but we want to mark it as parent of the next item, in case
+                # that next item is a list.
+                if use in exclude:
+                    # When ``use`` is in ``originals``, we don’t want
+                    # to mark it as parent of the next item because we want
+                    # to display the next item as root, otherwise the first
+                    # level of deleted objects would be shown nested.
+                    if use in originals:
```

Comment:
@BertrandBordage Would it be valid to set `parent_use = None` at this point? I think this is what's hurting my head so much about this code 😀 : the fact that `parent_use` is being left at its previous value in this case, which might be something left over from way earlier in the iteration. I suspect you're relying on an unspoken rule about the behaviour of `NestedObjects` which ensures that `parent_use` is only ever `None` at this point: i.e. it will never give you a situation like

    originals = exclude = {a, b}
    nested_list = [a, c, b, [d]]

where a non-original (`c`) appears at the top level before all the originals have been output. If so, setting `parent_use = None` would make the intention clearer, and save us from having to figure out whether that rule is really true.

Reply:
Yes, I’m indeed relying on that order rule, as I mentioned in my previous message.
You’re totally right, we should explicitly add another `parent_use = None` here to make it extra clear.
I tried too much to do an optimal algorithm…

Reply:
Sorry btw for the headache :frowning_face: